### PR TITLE
Add ctrl+z and fix undo for duration and table changes

### PR DESCRIPTION
**Pull request by Roelemans**
Closes PM-385

## Description

Adds shortcut ctrl-z and ctrl-shift-z for undo/redo

## Changes

* Added shortcuts
* Fixed undo/redo when changing duration
* Fixed undo/redo when editing one of the joint tables

### CODE REVIEW

**Comment by gaiavdh** [comment]
Have tried it and it works! Seems not to break anything.

File: march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_view.py [hunk 1]
```diff
@@ -36,6 +37,12 @@ def __init__(self):
 
         self.gait_type_combo_box.addItems(['walk_like', 'sit_like', 'stairs_like'])
 
+        self.initialize_shortcuts(self)
+
+    def initialize_shortcuts(self, parent):
+        self.ctrl_z = QShortcut(QKeySequence(self.tr('Ctrl+Z')), parent)
+        self.ctrl_shift_z = QShortcut(QKeySequence(self.tr('Ctrl+Shift+Z')), parent)
```
Comment (Olavhaasie) [inline march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_view.py]: this method has a `parent` argument, but you never call it with that argument. Why is that?

Comment (Olavhaasie) [inline march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_view.py]: I think it is better to name these shortcuts `undo_shortcut` and `redo_shortcut` then it is more clear elsewhere in the code what they do and if you want to change the key sequence you don't have to rename your variables.

Comment (Roelemans) [reply march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_view.py]: Woops that is an artefact of something I tried that did not work :sweat_smile:

Comment (Roelemans) [reply march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_view.py]: I thought the other way around - this way the view only knows that ctrl-z is a shortcut and not what it does. If you want to change what ctrl-z does, you don't have to change anything in the view. Its kind of the controller knowing about the view or the view knowing about the controller question, so I would prefer to keep it like this.

**Comment by Roelemans** [comment]
> For the most part this works nice, but when I add some setpoints, then change the gait name and then undo a couple of times, it will not undo further than the name change. Is this what you expected, because it can undo multiple setpoints.

I do not really understand what you explain here :sweat_smile:, I tried to reproduce the problem you describe but I could not :/. Can you be a bit more specific?

**Comment by RutgerVanBeek** [comment]
> > For the most part this works nice, but when I add some setpoints, then change the gait name and then undo a couple of times, it will not undo further than the name change. Is this what you expected, because it can undo multiple setpoints.
> 
> I do not really understand what you explain here , I tried to reproduce the problem you describe but I could not :/. Can you be a bit more specific?

So when you add a few set points, you can undo them all one by one, by using ctrl+z or the undo button. When I add a few set points and then change the gait name for instance the gait name, I can do ctrl+z to revert the changes in the gait name and not with the undo button. I cannot remove the setpoints I added before that with ctrl+z only with undo. It is probably a focus issue. It is only a minor issue so we could decide to just live with it, but the behavior of ctrl+z and undo differs sometimes.

**Comment by Roelemans** [comment]
Ahh I see. This is indeed a focus issue: the text editors have their own undo functionality. This can be nice since it allows undo/redo in text you type. Clicking next to a graph allows you to refocus on the gait generator. Arguably one could say this is a feature :innocent:. You probably won't edit the text boxes that often anyways.

**Pull request merged**